nexus/src/db/datastore/snapshot.rs

```diff
@@ -8,66 +8,62 @@ use super::DataStore;
 use crate::authz;
 use crate::context::OpContext;
 use crate::db;
-use crate::db::datastore::RunnableQuery;
+use crate::db::collection_insert::AsyncInsertError;
+use crate::db::collection_insert::DatastoreCollection;
 use crate::db::error::public_error_from_diesel_pool;
 use crate::db::error::ErrorHandler;
-use crate::db::lookup::LookupPath;
 use crate::db::model::Generation;
 use crate::db::model::Name;
+use crate::db::model::Project;
 use crate::db::model::Snapshot;
 use crate::db::model::SnapshotState;
 use crate::db::pagination::paginated;
 use crate::db::update_and_check::UpdateAndCheck;
 use async_bb8_diesel::AsyncRunQueryDsl;
 use chrono::Utc;
 use diesel::prelude::*;
-use nexus_types::identity::Resource;
 use omicron_common::api::external::CreateResult;
 use omicron_common::api::external::DataPageParams;
 use omicron_common::api::external::Error;
 use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupType;
+use omicron_common::api::external::ResourceType;
 use omicron_common::api::external::UpdateResult;
 use omicron_common::bail_unless;
 use uuid::Uuid;
 
 impl DataStore {
-    pub(super) async fn snapshot_ensure_query(
-        opctx: &OpContext,
-        authz_silo: &authz::Silo,
-        snapshot: Snapshot,
-    ) -> Result<impl RunnableQuery<Snapshot>, Error> {
-        opctx.authorize(authz::Action::CreateChild, authz_silo).await?;
-
-        use db::schema::snapshot::dsl;
-        Ok(diesel::insert_into(dsl::snapshot)
-            .values(snapshot)
-            .on_conflict((dsl::project_id, dsl::name))
-            .filter_target(dsl::time_deleted.is_null())
-            .do_nothing()
-            .returning(Snapshot::as_returning()))
-    }
-
     pub async fn project_ensure_snapshot(
         &self,
         opctx: &OpContext,
         authz_silo: &authz::Silo,
         snapshot: Snapshot,
     ) -> CreateResult<Snapshot> {
-        let snapshot_id = snapshot.id();
         let gen = snapshot.gen;
+        opctx.authorize(authz::Action::CreateChild, authz_silo).await?;
 
-        DataStore::snapshot_ensure_query(opctx, authz_silo, snapshot)
-            .await?
-            .execute_async(self.pool_authorized(opctx).await?)
-            .await
-            .map_err(|e| {
+        use db::schema::snapshot::dsl;
+        let project_id = snapshot.project_id;
+        let snapshot: Snapshot = Project::insert_resource(
+            project_id,
+            diesel::insert_into(dsl::snapshot)
+                .values(snapshot)
+                .on_conflict((dsl::project_id, dsl::name))
+                .filter_target(dsl::time_deleted.is_null())
+                .do_update()
+                .set(dsl::time_modified.eq(Utc::now())),
```

Comment:
This is one example case of https://github.com/oxidecomputer/omicron/issues/1168 .

The old code issued a query to ensure the snapshot exists, and then proceeded to perform a lookup to find the snapshot -- two queries instead of one.

Reply:
Did you mean this to set the time modified to itself rather than `now()`?  It seems like this will adjust the modification time even though no modification has been made.